eclair-core/src/main/scala/fr/acinq/eclair/payment/send/MultiPartPaymentLifecycle.scala

```diff
@@ -208,40 +207,50 @@ class MultiPartPaymentLifecycle(nodeParams: NodeParams, cfg: SendPaymentConfig,
       case Upstream.Local(_) => Upstream.Local(childId)
       case _ => cfg.upstream
     }
-    val childCfg = cfg.copy(id = childId, publishEvent = false, upstream = upstream)
+    val childCfg = cfg.copy(id = childId, publishEvent = false, recordMetrics = false, upstream = upstream)
     paymentFactory.spawnOutgoingPayment(context, childCfg)
   }
 
   private def gotoAbortedOrStop(d: PaymentAborted): State = {
     if (d.pending.isEmpty) {
-      myStop(d.request.replyTo, Left(PaymentFailed(id, paymentHash, d.failures)))
+      myStop(d.request, Left(PaymentFailed(id, paymentHash, d.failures)))
     } else
       goto(PAYMENT_ABORTED) using d
   }
 
   private def gotoSucceededOrStop(d: PaymentSucceeded): State = {
     d.request.replyTo ! PreimageReceived(paymentHash, d.preimage)
     if (d.pending.isEmpty) {
-      myStop(d.request.replyTo, Right(cfg.createPaymentSent(d.preimage, d.parts)))
+      myStop(d.request, Right(cfg.createPaymentSent(d.preimage, d.parts)))
     } else
       goto(PAYMENT_SUCCEEDED) using d
   }
 
-  def myStop(origin: ActorRef, event: Either[PaymentFailed, PaymentSent]): State = {
+  def myStop(request: SendMultiPartPayment, event: Either[PaymentFailed, PaymentSent]): State = {
     event match {
       case Left(paymentFailed) =>
         log.warning("multi-part payment failed")
-        reply(origin, paymentFailed)
+        reply(request.replyTo, paymentFailed)
       case Right(paymentSent) =>
         log.info("multi-part payment succeeded")
-        reply(origin, paymentSent)
+        reply(request.replyTo, paymentSent)
     }
-    Metrics.SentPaymentDuration
-      .withTag(Tags.MultiPart, Tags.MultiPartType.Parent)
-      .withTag(Tags.Success, value = event.isRight)
-      .record(System.currentTimeMillis - start, TimeUnit.MILLISECONDS)
-    if (retriedFailedChannels) {
-      Metrics.RetryFailedChannelsResult.withTag(Tags.Success, event.isRight).increment()
+    if (cfg.recordMetrics) {
+      val fees = event match {
+        case Left(paymentFailed) => 0 msat
+        case Right(paymentSent) => paymentSent.feesPaid
+      }
+      val success = event.isRight
+      val now = System.currentTimeMillis
+      val duration = now - start
+      context.system.eventStream.publish(PathFindingExperimentMetrics(cfg.recipientAmount, fees, success, duration, now, request.routeParams.experimentName))
```

Comment:
In the end the user doesn't care what kind of failure happened. The reason I've chosen these fields and not others is that it's all the end user can see: Did it work? How much did it cost? How long did it take?
We could add metrics but I'm not sure they would be useful. The goal of AB testing is not to help us debug individual payments but to tell us if on average a given set of parameters is better than another.

Reply:
I would agree if we had a lot of volume with a nice amount distribution, but that's not the case.

I'm afraid that just looking at the overall "it failed / it succeeded" signal without more information will make us take very misguided decisions because we would compare two experiments based on data that's not really comparable. I'm pretty sure that if we do just that, we will mostly do this loop:

- "Cool, experiment B has much better results than experiment A, let's switch it to use more than 50%"
- "Damn now it's experiment A that's better, let's switch back to experiment A using more than 50%"
- Rinse & Repeat

And in the meantime users would see failed payments that could be avoided...

Reply:
I don't know if that's unfortunate or not, but I see this A/B testing mechanism useful in the short-term *with manual work afterwards* if we have enough data in the DB. Yes it's not ideal, but I really think that's the reality of the data sets we'll work with for now.

Reply:
I've added `isMultiPart`. I don't think I can add the different types of failures: because of retries a single payment can fail in several different ways, how would you deal with that?

Reply:
Unless I'm missing something, it should be pretty simple to add with the way you currently emit the events since in `MultiPartPaymentLifecycle` you set `emitMetrics = false` to the child `PaymentLifecycle` and only emit the event when the payment is complete (failed or succeeded) after all retries have happened.

We need to do the following:

- When the payment failed, we should populate the `fees` field with the fee budget (instead of setting it to `0 msat`), which you can easily do using `routeParams.getMaxFee(recipientAmount)`
- When the payment failed, we should check all the `PaymentFailure` objects it contains, and if any of these comes from the final recipient (which you know by checking if it's a `RemoteFailure` where `e.originNode` matches the recipient) we should **not** emit the event

I think that once we've done that, the data we'll have should be relevant.

Can you also add the `targetNodeId` to the `PathFindingMetrics`? I really think it can be useful, it doesn't cost much to add, we can ignore it if we don't care about it, but if for some reason we need it it will be available.

Reply:
Added all that.